DogTales/dog/player/PlayerController.cpp

```diff
@@ -0,0 +1,41 @@
+#include "PlayerController.hpp"
+
+namespace dog {
+
+PlayerController::PlayerController() {
+	key_map.insert(std::make_pair("move_x", 0.f));
+	key_map.insert(std::make_pair("move_y", 0.f));
+	key_map.insert(std::make_pair("jump", 0.f));
+}
+
+void PlayerController::tick(bave::Seconds dt, bave::App const& app) {
+
+	auto const& key_state = app.get_key_state();
+
+	bool left = key_state.is_pressed(bave::Key::eA) || key_state.is_pressed(bave::Key::eLeft);
+	bool right = key_state.is_pressed(bave::Key::eD) || key_state.is_pressed(bave::Key::eRight);
+	bool up = key_state.is_pressed(bave::Key::eW) || key_state.is_pressed(bave::Key::eUp);
+	bool down = key_state.is_pressed(bave::Key::eS) || key_state.is_pressed(bave::Key::eDown);
+
+	key_map["move_x"] = 0.f;
+	key_map["move_x"] = left && !right ? -1.f : key_map["move_x"];
+	key_map["move_x"] = right && !left ? 1.f : key_map["move_x"];
+	key_map["move_x"] = right && left ? 0.f : key_map["move_x"];
+
+	key_map["move_y"] = 0.f;
+	key_map["move_y"] = down && !up ? -1.f : key_map["move_y"];
+	key_map["move_y"] = up && !down ? 1.f : key_map["move_y"];
+	key_map["move_y"] = up && down ? 0.f : key_map["move_y"];
+
+	key_map["jump"] = key_state.is_pressed(bave::Key::eW) || key_state.is_pressed(bave::Key::eUp) ? 1.f : 0.f;
+}
```

Comment:
A few things to refactor here. 
1. Don't call `key_map["move_x"]` a dozen times - it will perform the hash, comparison, and lookup each time. Instead get a reference once and operate on the reference throughout the function.
2. The `left` / `right` logic is unnecessarily complicating things, instead just append the delta to a temp `float`, and clamp it between `-1.0f` and `1.0f` at the end. 
3. This is still pretty hard-coded and there's no indirection for which key corresponds to which game-action. I think at this level it's better for the logic to be generic: the controller should cycle through its mappings and update the `key_map` accordingly, without knowing how many mappings there are / what they are.